javascript/ql/src/Statements/UseOfReturnlessFunction.ql

```diff
@@ -0,0 +1,103 @@
+/**
+ * @name Use of returnless function.
+ * @description Using the return value of a function that does not return anything is highly suspicious.
+ * @kind problem
+ * @problem.severity recommendation
+ * @id js/use-of-returnless-function
+ * @tags maintainability,
+ *       correctness
+ * @precision high
+ */
+
+import javascript
+import Declarations.UnusedVariable
+import Expressions.ExprHasNoEffect
+
+predicate returnsVoid(Function f) {
+  exists(f.getBody().(Stmt)) and
+  not f instanceof ExternalDecl and
+  not f.isGenerator() and
+  not f.isAsync() and
+  not exists(f.getAReturnedExpr())
+}
+
+predicate isStub(Function f) {
+    f.getBodyStmt(0) instanceof ThrowStmt
+    or
+    f.getBody().(BlockStmt).getNumChild() = 0
+}
+
+predicate benignContext(Expr e) {
+  inVoidContext(e) or 
+  
+  // A return statement is often used to just end the function.
+  exists(ReturnStmt ret | 
+    ret.getExpr() = e
+  )
+  or 
+  exists(ConditionalExpr cond | cond.getABranch() = e and benignContext(cond)) 
+  or 
+  exists(BinaryExpr bin | (bin.getOperator() = "&&" or bin.getOperator() = "||") and bin.getAnOperand() = e and benignContext(bin)) 
+  or
+  exists(SeqExpr parent | parent.getAnOperand() = e and benignContext(parent))
+  or
+  exists(ParExpr par | par.getExpression() = e and benignContext(par))
+  or
+  
+  // It is ok (or to be flagged by another query?) to await a non-async function. 
+  exists(AwaitExpr await | await.getOperand() = e) 
+  or
+  
+  // Avoid double reporting. It will always evaluate to false.
+  exists(IfStmt ifStmt | ifStmt.getCondition() = e)
```

Comment:
It may not be relevant anymore (if we move the alert location): but you should have a look athttps://github.com/Semmle/ql/blob/e048207e2f36b712e9d5e6fd42bf2639e17132f5/javascript/ql/src/Statements/UselessConditional.ql#L126-L152 to see how we support more that just `if`-statements.